### PR TITLE
expression: add EvalDuration/EvalTime for expression

### PR DESCRIPTION
**Pull request by XuHuaiyu**
BuiltinFunc like `addtime(datetime/time, time)` is different from `plus(datetime, time)`.

In `addtime`, we add two args as temporal type;
in `plus`, we add two args as int type.

It will be simpler if we evaluate using type Date/Duration directly 
than use string/numeric representation of Date/Datetime/Timestamp/Duration in some built-in functions.

PTAL @shenli @hanfei1991 @zimulala

### CODE REVIEW

**Comment by hanfei1991** [comment]
Do we need baseDateFunction and baseDurationFunction ?

**Comment by XuHuaiyu** [comment]
yeah, some built-in functions will return Datetime or Duration values. @hanfei1991

**Comment by XuHuaiyu** [comment]
PTAL @shenli @zimulala @hanfei1991

**Comment by hanfei1991** [comment]
LGTM

File: expression/expression.go [hunk 1]
```diff
@@ -66,6 +66,12 @@ type Expression interface {
 	// EvalDecimal returns the decimal representation of expression.
 	EvalDecimal(row []types.Datum, sc *variable.StatementContext) (val *types.MyDecimal, isNull bool, err error)
 
+	// EvalDate returns the date/datetime representation of expression.
+	EvalDate(row []types.Datum, sc *variable.StatementContext) (val types.Time, isNull bool, err error)
```
Comment (shenli) [inline expression/expression.go]: EvalTime is better, they are all MysqlTime types.

**Comment by XuHuaiyu** [comment]
PTAL @shenli

**Comment by shenli** [comment]
LGTM

**Pull request merged**